src/resolve/remote.rs

```diff
@@ -80,56 +93,64 @@ impl Remote {
         })
     }
 
-    fn content_type(&self, response: &Response) -> Option<ContentType> {
-        // Get a value content type header if any.
-        let header = response.headers().get(header::CONTENT_TYPE)
+    /// Injectively map urls to paths in the download directory.
+    fn target_path(&self, key: &FileKey) -> PathBuf {
+        let extension = key.path.extension();
+
+        let file_hash = Sha256::new()
+            .chain(key.path.to_str().unwrap())
+            .chain(key.mime.as_ref().map(Mime::as_ref).unwrap_or(""))
+            .result();
+
+        let mut hash_name = String::new();
+        file_hash.into_iter().for_each(|x| {
+            hash_name.push(std::char::from_digit((x >> 4).into(), 16).unwrap());
+            hash_name.push(std::char::from_digit((x & 15).into(), 16).unwrap());
+        });
+
+        if let Some(extension) = extension {
+            hash_name.push('.');
+            hash_name.push_str(extension.to_str().unwrap());
+        }
+
+        let mut target = self.temp.clone();
+        target.push(&key.host);
+        target.push(hash_name);
+
+        target
+    }
+}
+
+impl FileKey {
+    fn from(response: &Response) -> Self {
+        let mime = response.headers().get(header::CONTENT_TYPE)
             .and_then(|raw| raw.to_str().ok())
-            .and_then(|string| string.parse::<mime::Mime>().ok())?;
+            .and_then(|string| string.parse::<Mime>().ok());
 
-        match (header.type_(), header.subtype()) {
-            (mime::TEXT, sub) if sub == "markdown" => Some(ContentType::Markdown),
-            (mime::IMAGE, mime::PNG) | (mime::IMAGE, mime::JPEG) => Some(ContentType::Image),
-            (mime::APPLICATION, sub) if sub == "pdf" => Some(ContentType::Pdf),
-            // Let the file extension logic take over.
-            _ => None,
-        }
+        Self::from_parts(response.url().clone(), mime)
```

Comment:
s/.clone()/.to_owned()/
That makes it clearer that you have a reference here from which you want the owned version. With `.clone()` I was just wondering why you'd clone something in the very last line of a function.

Reply:
Instead we can make `from_parts` take `&Url`. All relevant components are cloned individually.